src/database/UIDatabase.js

```diff
@@ -35,36 +33,53 @@ export class UIDatabase {
     this.database = database;
   }
 
+  ERRORS = {
+    ERROR_IN_TRANSACTION: { code: 'ERROR_IN_TRANSACTION', message: 'Database is in a transaction' },
+    ERROR_NO_PERMISSION: { code: 'ERROR_NO_PERMISSION', message: 'Storage permission not granted' },
+    ERROR_UNKNOWN: { code: 'ERROR_UNKOWN', message: 'ERROR_UNKNOWN' },
+  };
+
   /**
-   * Closes the database, exports the realm file to '/Download/mSupplyMobile\ data' on device
-   * file system. The app will crash if anything tries to access the database while it is closed.
+   * Exports the realm file to '/Download/mSupplyMobile\ data' on device file system.
+   * Ensures there is enough space, the realm exists and requests storage permission,
+   * if required.
    */
-  exportData(filename = 'msupply-mobile-data') {
-    // TO DO: add method to react-native-database.
 
+  async exportData(filename = 'msupply-mobile-data') {
     const { realm } = this.database;
-    const realmPath = realm.path;
+    const { path: realmPath } = realm;
     const exportFolder = `${RNFS.ExternalStorageDirectoryPath}/Download/mSupplyMobile_data`;
+    const copyFileName = `${filename}${formatDate(new Date(), 'dashes')}`.replace(
+      /[~\\\\/|?*<:>"+]/g,
+      ''
+    );
+
+    const permissionParameters = {
+      permissionType: 'WRITE_EXTERNAL_STORAGE',
+      message: 'Export database',
+    };
+
+    // Before requesting permissions, ensure there is enough space and the realm
+    // file exists.
+    const exportValidation = await backupValidation(realmPath);
+    const { success } = exportValidation;
+    if (!success) return exportValidation;
+    // Request permissions for external storage before trying to backup the realm file.
+    const { success: permissionSuccess } = await requestPermission(permissionParameters);
+    // If permission was granted, ensure the realm is not in a transaction and backup
+    // the realm file.
+    if (!permissionSuccess) return { success: false, ...this.ERRORS.ERROR_NO_PERMISSION };
+    if (realm.isInTransaction) return { success: false, ...this.ERRORS.ERROR_IN_TRANSACTION };
+
+    // Finally try to create the backup/exported realm
+    try {
+      await RNFS.mkdir(exportFolder);
+      await RNFS.copyFile(realmPath, `${exportFolder}/${copyFileName}.realm`);
+    } catch (error) {
+      return { success: false, ...this.ERRORS.ERROR_UNKNOWN };
```

Comment:
any value in adding `error` to the message property?

Reply:
definitely, cheers ff2930f8d4232831b656fd516710f2c6025548bd